### ISSUE TITLE
SPIKE: automate agency verification process

### ISSUE THREAD

**Issue by stacysealky** [donategifts/donategifts#625]
Discussion session coming soon

validate EIN with CharityAPI
https://docs.charityapi.org/#introduction

**Comment by stacysealky** [comment]
Instead of doing manual agency verification, we should automate the process by 
- validating the EIN with CharityAPI 
- validating agency name, agency address, agency phone, agency website(optional) with Google Maps API or SerpAPI
- email verification of user using work email address that contains "@agency-website" 

When these 3 checks pass, then automatically set isVerified to true 
and when agency.isVerified, send an automated email 

What do you guys think? @Enubia @nick-w-nick

**Comment by Enubia** [comment]
Can we actually fill in the agency information with what's provided by the API?

This way they wouldn't need to provide the address and things like that

**Comment by stacysealky** [comment]
> Can we actually fill in the agency information with what's provided by the API?
> 
> This way they wouldn't need to provide the address and things like that
> 
> ```json
> {
>     "city": "city",
>     "ein": "Employer Identification Number, or Tax ID Number",
>     "name": "The name of the organization",
>     "organization": "  This defines the type of organization as follows:\n        Code Description\n        1 Corporation\n        2 Trust\n        3 Co-operative\n        4 Partnership\n        5 Association",
>     "state": "state",
>     "street": "Street Address Line 1",
>     "zip": "zip code"
> }
> ```

Hmmm, I don't like the idea of filling in the verification information for them. 
It feels like an inappropriate choice for security and privacy compliances. 
I would prefer the users to enter all of their info and then we verify what they provided is legitimate.

But that's just my thought. I want to understand why you advocate this method @Enubia 
Any suggestions? @nick-w-nick

**Comment by Tokikko** [comment]
IS it not better to automate this with an api as Marco said? The data should be correct (i assume the api has correct data) and they have just to enter their EIN number and be done. After that we can take them to a profile page where they can edit some of the data if they need to. Seems that would be much easier for them since its only 1 filed they have to fill and there is less room for them to make mistakes when filling the form.

Why would it be "inappropriate choice for security and privacy compliances"?

**Comment by Enubia** [comment]
Well, it would be the same as with Google address autofill, wouldn't it? This can also autofill your data, chrome also can autofill your saved data.

I'd rather have my data prefilled and just need to check it instead of manually inputting it. The API was made to have this data available. If they provide us their EIN already then there's no reason for security concerns. We just use what the API provides and as @Tokikko already said
> After that we can take them to a profile page where they can edit some of the data if they need to.

We can also show a confirmation screen where we tell them that this data was fetched from the charity API and that they need to verify the correctness of the prefilled data.

**Comment by nick-w-nick** [comment]
> > Can we actually fill in the agency information with what's provided by the API?
> > This way they wouldn't need to provide the address and things like that
> > ```json
> > {
> >     "city": "city",
> >     "ein": "Employer Identification Number, or Tax ID Number",
> >     "name": "The name of the organization",
> >     "organization": "  This defines the type of organization as follows:\n        Code Description\n        1 Corporation\n        2 Trust\n        3 Co-operative\n        4 Partnership\n        5 Association",
> >     "state": "state",
> >     "street": "Street Address Line 1",
> >     "zip": "zip code"
> > }
> > ```
> 
> Hmmm, I don't like the idea of filling in the verification information for them. It feels like an inappropriate choice for security and privacy compliances. I would prefer the users to enter all of their info and then we verify what they provided is legitimate.
> 
> But that's just my thought. I want to understand why you advocate this method @Enubia Any suggestions? @nick-w-nick

While we could verify address information during agency verification, there are quite a few edge cases that can arise from trying to validate two addresses against each other due to formatting issues and user input.

We could just have the agency entire their agency name and EIN first, fetch the data from the CharityAPI, validate the EIN and name matches, and then display the address from the API as the suggested address to the user and they can verify if the address is correct or they can supply an updated one if needed.

If we definitely want to have the agency provide its address in entirety before reaching out to the CharityAPI, we should validate everything except for the street address.

**Comment by stacysealky** [comment]
Yeah let's think more about it. I appreciate all the above suggestions, btw. 

One point I'd like to make: 
Having agencies enter their 'agency name' and 'agency EIN' alone will not be sufficient for verification because people can look these up and create fake accounts. 
We must have them enter their work email that ends with "@agencywebsite". 
And then have them click the verification link through that email, then we can set agency.isVerified.

So if we do the prefill: 
1. Agency partner needs to go through the first step of entering their work email 
See #570 

2. Agency partner enters their agency name, agency EIN, agency website, and agency bio.
We need to make agency website a requirement if we do the automated verification, since we will be sending the link to email@agencywebsite. 
Agency bio has to be a manual entry text area. 
Agency website, I think, has to be manual entry too, because I'm not sure if they registered that information with IRS or GoogleMaps, so it may or may not contain that data. 

3. If EIN exists and matches the agency name, 
then in the next step we show the editable pre-filled values for address line 1, address line 2, city, zip, state, country, and phone.

4. User submits, then clicks the verification link sent to their email@agencywebsite.

5. Agency is verified.


So in conclusion, I'm fine with the prefill but only for address line 1, address line 2, city, zip, state, country, and phone.
Definitely don't prefill the EIN. That is our first verification checkpoint.

**Comment by stacysealky** [comment]
> Why would it be "inappropriate choice for security and privacy compliances"?

@Tokikko because according to Marco's attached JSON, we will be prefilling EIN also. That is not good. We should only prefill address line 1, address line 2, city, zip, state, country, and phone.

**Comment by Tokikko** [comment]
Hm maybe im reading it wrong but users have to fill EIN based on that input we fetch the agency details (https://api.charityapi.org/api/organizations/:ein). 

Since we need the email to confirm agency we can keep the current 1 page of signup and on the second instead of the agency form just ask for EIN fetch the data for it and then show them the filled form. They can adjust it then and fill anything else that is necessary.

**Comment by Enubia** [comment]
> 
> 
> > Why would it be "inappropriate choice for security and privacy compliances"?
> 
> 
> 
> @Tokikko because according to Marco's attached JSON, we will be prefilling EIN also. That is not good. We should only prefill address line 1, address line 2, city, zip, state, country, and phone. 

I never mentioned pre filling the EIN. What I wanted to be prefilled was the Address information since we need the EIN to actually get the data. The JSON EIN number is only there to verify that the user did in fact provide us the correct number for his agency.

**Comment by stacysealky** [comment]
> > > Why would it be "inappropriate choice for security and privacy compliances"?
> > 
> > 
> > @Tokikko because according to Marco's attached JSON, we will be prefilling EIN also. That is not good. We should only prefill address line 1, address line 2, city, zip, state, country, and phone.
> 
> I never mentioned pre filling the EIN. What I wanted to be prefilled was the Address information since we need the EIN to actually get the data. The JSON EIN number is only there to verify that the user did in fact provide us the correct number for his agency.

Oh ok, I thought you implied prefilling the JSON object you attached. And I saw the EIN there and thought we shouldn't.
Either way, yeah we are on the same page. I think we should prefill only the address and phone. And even for those, make them immediately editable, so they can correct something if incorrect/changed.
